Adding GobblinTestEventBusWriter

chavdar opened the pull request:
Adding some instrumentation to allow us to test end-to-end gobblin jobs.

The idea is to write output to a Guava EventBus and then have some utilities to assert the output.

Changes:
- TestingEventBuses - maintains a global list of testing event buses. This is to simplify the sharing of buses between the writer(s) and the asserter(s).
- GobblinTestEventBusWriter - a writer implementation to write to an EventBus managed by TestingEventBuses
- TestingEventBusAsserter - helper functions to do asserts on the incoming events from TestingEventBuses.
- Unit tests.

@ibuenros @abti can you review?

abti commented on gobblin-core/src/main/java/gobblin/writer/test/TestingEventBusAsserter.java:
```diff
@@ -0,0 +1,169 @@
+        if (! remainingExpectedValues.contains(input)) {
+          return false;
+        }
+        remainingExpectedValues.remove(input);
```
What would happen if two duplicate values are produced / asserted?

chavdar replied:
This specific asset implementation will fail. It is meant to check only for permutations of distinct values.

chavdar commented:
Any more comments?

ibuenros commented on gobblin-core/src/main/java/gobblin/writer/test/TestingEventBusAsserter.java:
```diff
@@ -0,0 +1,169 @@
+
+  @AllArgsConstructor
+  public static class StaticMessage implements Function<TestingEventBuses.Event, String> {
+    private final String errorMessage;
```
Can you rename this to just `message`?

chavdar replied:
Fixed.

chavdar commented:
@ibuenros 

I have added a sample usage. See
- SimpleHelloWorldJob.jobconf for a sample configuration for using GobblinTestEventBusWriter
- TestStandardGobblinInstanceLauncher.testSubmitToJobCatalog() for refactoring of an old method to validate the output of the job.

I have also made some small improvements:
- writer.GobblinTestEventBusWriter.eventBusId config to set explicitly the eventbus id because using the writer output path is complicated.
- Added a helper method to HelloWorldSource to simplify validation of the output.
